File: app/src/main/java/ru/igla/tfprofiler/core/tflite/TensorFlowUtils.java
package ru.igla.tfprofiler.core.tflite;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.content.res.AssetManager;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.tensorflow.lite.support.metadata.MetadataExtractor;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.List;

import ru.igla.tfprofiler.models_list.domain.ModelEntity;
import ru.igla.tfprofiler.utils.IOUtils;
import timber.log.Timber;


public final class TensorFlowUtils {

    private TensorFlowUtils() {
    }

    /***
     * https://www.tensorflow.org/lite/performance/nnapi#create_a_device_exclusion_list
     * In production, there may be cases where NNAPI does not perform as expected.
     * We recommend developers maintain a list of devices that should not use NNAPI acceleration in combination with particular models.
     * @return board platform
     */
    public static String getBoardPlatform() {
        return getProp("ro.board.platform");
    }

    public static String getProductDevice() {
        return getProp("ro.product.device");
    }

    public static String getProp(String propName) {
        String boardPlatform = "";
        try {
            Process sysProcess =
                    new ProcessBuilder("/system/bin/getprop", propName).
                            redirectErrorStream(true).start();
            try (BufferedReader reader = new BufferedReader
                    (new InputStreamReader(sysProcess.getInputStream()))) {
                String currentLine;
                while ((currentLine = reader.readLine()) != null) {
                    boardPlatform = currentLine;
                }
            }
            sysProcess.destroy();
        } catch (IOException e) {
            Timber.e(e);
        }
        Log.d("Board Platform", boardPlatform);
        return boardPlatform;
    }

    @NonNull
    public static List<String> loadLabelList(@NonNull AssetManager assetManager, String labelPath) throws IOException {
        try (InputStream is = assetManager.open(labelPath)) {
            return loadLabelList(is);
        }
    }

    @NonNull
    public static List<String> loadLabelList(@NonNull InputStream inputStream) throws IOException {
        List<String> labelList = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream))) {
            String line;
            while ((line = reader.readLine()) != null) {
                Timber.w(line);
                labelList.add(line);
            }
        }
        return labelList;
    }

    /**
     * Memory-map the model file in Assets.
     * Preload and memory map the model file, returning a MappedByteBuffer containing the model.
     */
    public static MappedByteBuffer loadModelFileFromAssets(Context context, String modelFilename)
            throws IOException {
        AssetManager assetManager = context.getAssets();
        try (AssetFileDescriptor fileDescriptor = assetManager.openFd(modelFilename)) {
            try (FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor())) {
                FileChannel fileChannel = inputStream.getChannel();
                long startOffset = fileDescriptor.getStartOffset();
                long declaredLength = fileDescriptor.getDeclaredLength();
                return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
            }
        }
    }

    public static MappedByteBuffer loadModelFileFromExternal(String modelFilename)
            throws IOException {
        File file = new File(modelFilename);
        try (FileInputStream inputStream = new FileInputStream(file)) {
            FileChannel fileChannel = inputStream.getChannel();
            return fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, file.length());
        }
    }

    public static boolean isAssetFileExists(Context context, String filename) {
        AssetManager mg = context.getResources().getAssets();
        InputStream is = null;
        try {
            is = mg.open(filename);
            return true;
        } catch (IOException ex) {
            //ignore
            return false;
        } finally {
            IOUtils.closeQuietly(is);
        }
    }

    @Nullable
    public static MetadataExtractor getMetaData(@NonNull Context context, @NonNull String modelFilename) {
        try {
            MappedByteBuffer m = loadModelFileFromAssets(context, modelFilename);
            MetadataExtractor metadataExtractor;
            metadataExtractor = new MetadataExtractor(m);
            if (!metadataExtractor.hasMetadata()) {
                return null;
            }
            return metadataExtractor;
        } catch (IOException e) {
            Timber.e(e);
        }
        return null;
    }

    public static long getModelFileSize(@NonNull Context context, @NonNull ModelEntity modelEntity) {
        if (modelEntity.getModelType().isCustomModel()) {
            return new File(modelEntity.getModelFile()).length();
        }
        try {
            String modelFilename = modelEntity.getModelFile();
            try (AssetFileDescriptor fd = context.getAssets().openFd(modelFilename)) {
                return fd.getLength();
            }
        } catch (IOException e) {
            Timber.e(e);
            return AssetFileDescriptor.UNKNOWN_LENGTH;
        }
    }

    public static float sigmoid(final float x) {
        return (float) (1. / (1. + Math.exp(-x)));
    }
}

File: app/src/main/java/ru/igla/tfprofiler/core/tflite/TensorFlowUtils.kt
package ru.igla.tfprofiler.core.tflite

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.content.res.AssetManager
import org.tensorflow.lite.support.metadata.MetadataExtractor
import ru.igla.tfprofiler.models_list.domain.ModelEntity
import timber.log.Timber
import java.io.*
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.util.*
import kotlin.math.exp

object TensorFlowUtils {

    @Throws(IOException::class)
    fun loadLabelList(assetManager: AssetManager, labelPath: String): List<String> {
        assetManager.open(labelPath).use { ins -> return loadLabelList(ins) }
    }

    @Throws(IOException::class)
    fun loadLabelList(inputStream: InputStream): List<String> {
        val labelList: MutableList<String> = ArrayList()
        BufferedReader(InputStreamReader(inputStream)).use { reader ->
            var line: String
            while (reader.readLine().also { line = it } != null) {
                Timber.w(line)
                labelList.add(line)
            }
        }
        return labelList
    }

    /**
     * Memory-map the model file in Assets.
     * Preload and memory map the model file, returning a MappedByteBuffer containing the model.
     */
    @Throws(IOException::class)
    fun loadModelFileFromAssets(context: Context, modelFilename: String?): MappedByteBuffer {
        val assetManager = context.assets
        assetManager.openFd(modelFilename!!).use { fileDescriptor ->
            FileInputStream(fileDescriptor.fileDescriptor).use { inputStream ->
                val fileChannel = inputStream.channel
                val startOffset = fileDescriptor.startOffset
                val declaredLength = fileDescriptor.declaredLength
                return fileChannel.map(
                    FileChannel.MapMode.READ_ONLY,
                    startOffset,
                    declaredLength
                )
            }
        }
    }

    @Throws(IOException::class)
    fun loadModelFileFromExternal(modelFilename: String): MappedByteBuffer {
        val file = File(modelFilename)
        FileInputStream(file).use { inputStream ->
            val fileChannel = inputStream.channel
            return fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, file.length())
        }
    }

    fun isAssetFileExists(context: Context, filename: String): Boolean {
        val assetManager = context.resources.assets
        try {
            assetManager.open(filename).use {
                return true
            }
        } catch (e: Exception) {
            return false
        }
    }

    fun getMetaData(context: Context, modelFilename: String): MetadataExtractor? {
        try {
            val m = loadModelFileFromAssets(context, modelFilename)
            val metadataExtractor = MetadataExtractor(m)
            return if (!metadataExtractor.hasMetadata()) {
                null
            } else metadataExtractor
        } catch (e: IOException) {
            Timber.e(e)
        }
        return null
    }

    fun getModelFileSize(context: Context, modelEntity: ModelEntity): Long {
        if (modelEntity.modelType.isCustomModel()) {
            return File(modelEntity.modelFile).length()
        }
        try {
            val modelFilename = modelEntity.modelFile
            context.assets.openFd(modelFilename).use { fd -> return fd.length }
        } catch (e: IOException) {
            Timber.e(e)
            return AssetFileDescriptor.UNKNOWN_LENGTH
        }
    }

    fun sigmoid(x: Float): Float {
        return (1.0 / (1.0 + exp(-x.toDouble()))).toFloat()
    }
}